proposals/0024-repair-and-restart.md

```diff
@@ -0,0 +1,200 @@
+arguments.
+
+The current process involves a lot of human intervention, if people make a
+mistake in deciding the highest optimistically confirmed slot, it could mean
+rollback of user transactions after those transactions have been confirmed,
+which is not acceptable.
+
+We aim to automate the finding of highest optimistically confirmed slot and
+block data distribution, so that we can lower the possibility of human mistakes
+in the cluster restart process.
+
+## Alternatives Considered
+
+See [Handling Common Solana Outages](https://docs.google.com/document/d/1RkNAyz-5aKvv5FF44b8SoKifChKB705y5SdcEoqMPIc)
+for details.
+
+There are many proposals about automatically detecting that the cluster is
+in an outage so validators should enter a recovery process automatically.
+
+While getting human out of the loop greatly improves recovery speed,
+automaticlly restarting the whole cluster still seems risky. Because if
+the recovery process itself doesn't work, it might be some time before
+we can get human's attention. And it doesn't solve the cases where new binary
+is needed. So for now we still plan to have human in the loop.
+
+## Detailed Design
+
+The new protocol tries to make all 80% restarting validators get the same
+data blocks and the same set of last votes among them, then they can probably
+make the same decision and then proceed.
+
+The steps roughly look like this:
+
+1. Everyone freezes, no new blocks, no new votes, and no Turbine
+
+2. Make all blocks which can potentially have been optimistically confirmed
+before the freeze propagate to everyone
+
+3. Make restart participants' last votes before the freeze propagate to
+everyone
+
+4. Now see if enough people can optimistically agree on one block (same slot
+and hash) to restart from
+
+4.1 If yes, proceed and restart
+
+4.2 If no, freeze and print out what you think is wrong, wait for human
+
+A new command line arg --RepairAndRestart is added. When the cluster is in need
+of a restart, we assume at least 80% will restart with this arg. Any validators
+restarted with this arg does not participate in the normal Turbine protocol,
+update its vote, or generate new blocks until all of the following steps are
+completed.
+
+### Gossip last vote before the restart and ancestors on that fork
+
+Send Gossip message LastVotedForkSlots to everyone in restart, it contains the
+last voted slot on its tower and the ancestor slots on the last voted fork and
+is sent in a compressed bitmap like the EpicSlots data structure. The number of
+ancestor slots sent is hard coded at 81000, because that's 400ms * 81000 = 9
+hours, we assume most restart decisions to be made in 9 hours. If a validator
+restarts after 9 hours past the outage, it cannot join the restart this way. If
+enough validators failed to restart within 9 hours, then use the old restart
+method.
+
+The fields of LastVotedForkSlots are:
+
+- `last_voted_slot`: the slot last voted, this also serves as last_slot for the
```

Comment:
WDYT about starting at the `highest_oc_slot` instead of `last_voted_slot`? Now that we're back to pruning off everything above the highest oc slot anyway, it seems unnecessary to share votes beyond the oc slot and should simplify the `Heaviest`-collection logic since only the reported oc slots need to be considered

Reply:
Because locally you don't know what highest_oc_slot until you hear back from 80% of the peers what they voted?

The only thing you can make sure locally upon restart is what you personally voted last time.

Reply:
Each validator already records their local highest oc slot, so they can easily just regurgitate that during the restart though. If a global higher oc slot is then observed, that validator knows it needs to repair, and of course the higher global oc slot should be a sibling of its local highest oc slot else panic

Reply:
The global highest oc slot might not be anyone's local highest oc slot. I think we need to be careful here, since the definition of highest oc slot is that >2/3 voted for it, sending last vote is really the best bet. Since here every slot is only one bit, we don't have to be too optimal here, it's okay to send more slots.

Reply:
> The global highest oc slot might not be anyone's local highest oc slot.

 If it's not anyone's local highest oc slot then it was never observed, so does it really exist? This seems like an extreme edge case that's barely worth considering.

Reply:
For example if the restart protocol ever produced a global highest oc slot that nobody locally could produce with `solana-ledger-tool -l ledger latest-optimistic-slots`, I'd say that's most likely a bug or attack.

Reply:
Let me rephrase it, it could be whoever observed the highest oc slot is not in the restart, then users could see that their block got enough votes but somehow was rolled back.
I think it's important to guarantee to users that optimistically confirmed block will never be rolled back in all cases using some kind of proof in the algorithm.
And by guaranteeing the rigorous proof we are not losing much, it's just sending a few more slots in this message.

Reply:
> it could be whoever observed the highest oc slot is not in the restart,

hmm, but if 66%+ voted on that slot then is that really possible if we require 80% of the stake for a restart?

`LastVotedForkSlots` doesn't seem like a rigorous proof. Validators can lie about it with no consequence, specifically by claiming `last_voted_slot` is lower than what it truly was. Later there may be slashing, somehow, but that's not the case today or really dealt with in this SIMD.

Reply:
It's possible that 66% voted for it, but only 20% observed all the votes, the rest only observed part of the votes.

Okay, I don't have rigorous proof, but I do have something close to a proof, I can work on a formal proof if you really want that. You are right that validators can always lie in LastVotedForkSlots, but we do have assumptions that only 5% are non-conforming and they lie. And, if validators can lie about Last voted slot, they can also lie about local oc slot, I don't see a big difference here.

https://docs.google.com/document/d/1RkNAyz-5aKvv5FF44b8SoKifChKB705y5SdcEoqMPIc/edit#heading=h.ooavt3aiyxck

Note that: all the 47% etc numbers are calculated assuming we send last vote here, if we definitely need to replace last vote with locally oc slot, then all of these numbers stand on sand and we have to redesign the protocol.

I really think sending last vote is a relatively cheap price we pay for giving our users some assurance that optimistically confirmed slot is never rolled back. We could tell them local oc slot mostly works, but we couldn't really easily say under what percentage of cases this would work, and to an end-user that sounds a bit unassuring.

Reply:
Ok fair enough. Mostly just exploring the problem a little. 

How about something about how the validator should respond if the restart process derives a highest oc slot that's not equal to or a child of validator's local oc slot seems useful: probably panic and request a human

Reply:
Sure, we can send a panic if a validator finds out highest oc slot is not a child of local oc slot. Added to the last step.